### PR TITLE
[FE] Add New Pet

### PR DESCRIPTION
**Pull request by ShaunLWM**
Previously for Edit Pet, but it seemed like the design was changed, so I move this to Add Pet.
Since backend is not confirmed yet, I have not added in data validation and form submission.

### CODE REVIEW

**Comment by stonefruit** [comment]
hmm shouldn't Add Pet and Edit Pet be the same form? was it intended to have 2 different ones?

This page will probably be the most used page for the beginning at least, so we will surely need to provide decent validation. Perhaps you could add some boilerplate data validation so we can easily update once we confirm?

**Comment by ShaunLWM** [comment]
That was what I thought as well. I thought it would be the same design but it's not for some reason :/
https://www.figma.com/file/9nCpbCGXXTB7Aj3nHohEtB/Shelter-Side-Desktop-UI?node-id=152%3A31661

Will add the data validation once I'm free. Will keep this PR open for now. Plus I think we haven't confirm the model on the backend AFAIK. So I'm just afraid of double work again.

File: frontend/pages/shelter/pet/add.tsx [hunk 1]
```diff
@@ -0,0 +1,443 @@
...
+		if (!value) return;
+		setPet((prev) => ({ ...prev, [key]: value }));
+	};
+
+	const onDateChange = (date: moment.Moment, key: keyof Pick<PetData, "acquired" | "dateOfBirth">) => {
+		if (!date) return;
+		setPet((prev) => ({ ...prev, [key]: date.toDate() }));
+	};
+
+	const onGalleryChange = (images: string[]) => {
```
Comment (ryhuz) [inline frontend/pages/shelter/pet/add.tsx:57]: Getting an error due to this as it is passed in as a prop to `ImageGallery` and is then a dependency for `useEffect`

**Comment by stonefruit** [comment]
> That was what I thought as well. I thought it would be the same design but it's not for some reason :/ https://www.figma.com/file/9nCpbCGXXTB7Aj3nHohEtB/Shelter-Side-Desktop-UI?node-id=152%3A31661
> 
> Will add the data validation once I'm free. Will keep this PR open for now. Plus I think we haven't confirm the model on the backend AFAIK. So I'm just afraid of double work again.

@qroll is working on the endpoints for the animals, so when she's done can confirm the model

File: frontend/pages/shelter/pet/add.tsx (hunk 1 above)
Comment (ryhuz) [reply frontend/pages/shelter/pet/add.tsx:57]: actual error might be due to how the set of images is being updated repeatedly

File: frontend/pages/shelter/pet/add.tsx [hunk 2]
```diff
@@ -0,0 +1,443 @@
...
+			setCloned(prev => {
+				return [...prev, reader.result as string];
+			})
+		};
+
+		reader.readAsDataURL(pickedImage);
+	}, [pickedImage]);
+
+	useEffect(() => {
+		if (!onChange) return;
```
Comment (stonefruit) [inline frontend/pages/shelter/pet/add.tsx:285]: Following up on @ryhuz comment, `onChange` is always present. `ImageGalleryProps` should not have optional `onChange` as it is always defined.

File: frontend/pages/shelter/pet/add.tsx [hunk 3]
```diff
@@ -0,0 +1,443 @@
...
+import { FurLength, PetData, Sex, Species, Status } from "common/enums";
+import ShelterLayout from "layouts/shelter/ShelterLayout";
+import moment from "moment";
+import { useRouter } from "next/dist/client/router";
+import React, { ReactNode, useEffect, useRef, useState } from "react";
+import styled from "styled-components";
+
+export default function EditPetDetails() {
+	const router = useRouter();
+	const { id } = router.query;
```
Comment (ryhuz) [inline frontend/pages/shelter/pet/add.tsx:22]: this is currently unused. accident or future proof?

Comment (ShaunLWM) [reply frontend/pages/shelter/pet/add.tsx]: Forget to remove. Paiseh.

**Comment by ryhuz** [comment]
there is currently a refactor of shelter URL and fetching of shelter details and pets. might affect this branch

File: frontend/pages/shelter/pet/add.tsx [hunk 4]
```diff
@@ -0,0 +1,231 @@
...
+
+	const onDateChange = (
+		date: moment.Moment,
+		key: keyof Pick<PetData, "acquired" | "dateOfBirth">
+	) => {
+		if (!date) return;
+		setPet((prev) => ({ ...prev, [key]: date.toDate() }));
+	};
+
+	const updateGallery = useCallback((images: string[]) => {
```
Comment (stonefruit) [inline frontend/pages/shelter/pet/add.tsx:77]: what is the purpose of useCallback here?

**Review by stonefruit — COMMENTED** [review]
Looks good. Added a comment on why use `useCallback`. If there is not reason please remove to reduce code complexity. thx

File: frontend/pages/shelter/pet/add.tsx [hunk 5]
```diff
@@ -0,0 +1,443 @@
...
+	);
+}
+
+// =============================================================================
+// Secondary Components
+// =============================================================================
+
+type ImageGalleryProps = {
+	images?: string[];
+	onChange?: (images: string[]) => void;
```
[the remark this answers is not in the document]
Comment (ShaunLWM) [reply frontend/pages/shelter/pet/add.tsx]: The `onChange` callback is optional as I'm using it together with the `isEditMode`. If photo grid is in edit mode, any changes will trigger `onChange`, whereas in viewing mode, `onChange` isn't needed.

File: frontend/pages/shelter/pet/add.tsx (hunk 4 above)
Comment (ShaunLWM) [reply frontend/pages/shelter/pet/add.tsx:77]: To stop the recursive update. I'm not 100% sure on how to solve this actually - not sure if my implementation is correct. @ryhuz please help me take a look again leh 😥

**Comment by stonefruit** [comment]
@ShaunLWM Please take a look at comments and code for https://github.com/bettersg/PawScore/pull/44/files

**Comment by ShaunLWM** [comment]
I guess this can be merged as well?

**Review by stonefruit — APPROVED** [review]

**Pull request merged**